tools/quake3/common/vfs.c

```diff
@@ -188,38 +231,48 @@ void vfsInitDirectory( const char *path ){
 		dir = g_dir_open( path, 0, NULL );
 
 		if ( dir != NULL ) {
-			while ( 1 )
+
+			while(1)
 			{
-				const char* name = g_dir_read_name( dir );
-				if ( name == NULL ) {
+				const char* name = g_dir_read_name(dir);
+				if (name == NULL) {
 					break;
 				}
 
-				dirlist = g_strdup( name );
+				char* direntry = g_strdup(name);
+				dirlist = g_slist_append(dirlist, direntry);
+			}
 
-				{
-					char *ext = strrchr( dirlist, '.' );
+			dirlist = g_slist_sort(dirlist, vfsPakSort);
 
-					if ( ext != NULL && ( !Q_stricmp( ext, ".pk3dir" ) || !Q_stricmp( ext, ".dpkdir" ) ) ) {
-						if ( g_numDirs == VFS_MAXDIRS ) {
-							continue;
-						}
-						snprintf( g_strDirs[g_numDirs], PATH_MAX, "%s/%s", path, name );
-						g_strDirs[g_numDirs][PATH_MAX-1] = '\0';
-						vfsFixDOSName( g_strDirs[g_numDirs] );
-						vfsAddSlash( g_strDirs[g_numDirs] );
-						++g_numDirs;
-					}
+			while ( dirlist )
+			{
+				GSList *cur = dirlist;
+				char* name = (char*)cur->data;
+				if ( name == NULL ) {
+					break;
+				}
+
+				char *ext = strrchr(name, '.' );
 
-					if ( ext == NULL || ( Q_stricmp( ext, ".pk3" ) != 0 && Q_stricmp( ext, ".dpk" ) != 0 ) ) {
+				if ( ext != NULL && ( !Q_stricmp( ext, ".pk3dir" ) || !Q_stricmp( ext, ".dpkdir" ) ) ) {
+					if ( g_numDirs == VFS_MAXDIRS ) {
 						continue;
 					}
+					snprintf( g_strDirs[g_numDirs], PATH_MAX, "%s/%s", path, name );
+					g_strDirs[g_numDirs][PATH_MAX-1] = '\0';
+					vfsFixDOSName( g_strDirs[g_numDirs] );
+					vfsAddSlash( g_strDirs[g_numDirs] );
+					++g_numDirs;
 				}
 
-				sprintf( filename, "%s/%s", path, dirlist );
-				vfsInitPakFile( filename );
+				if ( ext != NULL && ( !Q_stricmp( ext, ".pk3" ) || !Q_stricmp( ext, ".dpk" )) ) {
+					sprintf(filename, "%s/%s", path, name);
+					vfsInitPakFile(filename);
+				}
 
-				g_free( dirlist );
+				g_free(name);
+				dirlist = g_slist_remove(cur, name);
```

Comment:
Is a little safer to free after the list remove, but that should be ok.

Reply:
Oops 😁 thanks for fixing that!